File: lab4mapreduce.java
import java.util.*;
import java.util.stream.*;

public class lab4mapreduce {

    public static void main(String[] args) {
        // Sample list of integers
        List<Integer> numbers = Arrays.asList(1, 2, 3, 4, 5, 6);

        // 1. Filter: Get only even numbers
        List<Integer> evenNumbers = numbers.stream()
                                           .filter(n -> n % 2 == 0) // Lambda for filter
                                           .collect(Collectors.toList());
        System.out.println("Even Numbers: " + evenNumbers);

        // 2. Map: Square the even numbers
        List<Integer> squaredEvens = evenNumbers.stream()
                                                .map(n -> n * n) // Lambda for map
                                                .collect(Collectors.toList());
        System.out.println("Squared Even Numbers: " + squaredEvens);

        // 3. Reduce: Sum of the squared even numbers
        Optional<Integer> sumOfSquares = squaredEvens.stream()
                                                     .reduce((a, b) -> a + b); // Lambda for reduce

        // 4. Optional Handling
        if (sumOfSquares.isPresent()) {
            System.out.println("Sum of Squared Even Numbers: " + sumOfSquares.get());
        } else {
            System.out.println("No squared even numbers to sum.");
        }

        // 5. Optional function example: Get first even number
        Optional<Integer> firstEven = getFirstEven(numbers);
        System.out.println("First Even Number: " + firstEven.orElse(-1));
    }

    // Function that returns Optional<Integer>
    public static Optional<Integer> getFirstEven(List<Integer> list) {
        return list.stream()
                   .filter(n -> n % 2 == 0)
                   .findFirst(); // returns Optional
    }
}

File: lab4mapreduce.py
from functools import reduce
from typing import Optional

numbers = [1, 2, 3, 4, 5, 6]

# Filter: Get only even numbers
even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
print("Even Numbers:", even_numbers)

# Map: Square each even number
squared_evens = list(map(lambda x: x ** 2, even_numbers))
print("Squared Even Numbers:", squared_evens)

# Reduce: Sum of squared even numbers
sum_of_squares: Optional[int] = reduce(lambda x, y: x + y, squared_evens) if squared_evens else None

# Optional Handling
if sum_of_squares is not None:
    print("Sum of Squared Even Numbers:", sum_of_squares)
else:
    print("No even numbers to process.")

# Demonstrate use of Optional in function return
def get_first_even(nums: list[int]) -> Optional[int]:
    """Return the first even number from the list, if any."""
    evens = list(filter(lambda x: x % 2 == 0, nums))
    return evens[0] if evens else None

first_even = get_first_even(numbers)
print("First Even Number:", first_even if first_even is not None else "No even number found.")
